pkg/providers/imagefamily/image.go

```diff
@@ -91,57 +86,116 @@ func (p *Provider) KubeServerVersion(ctx context.Context) (string, error) {
+			return p.GetLatestImageID(ctx, defaultImage, nodeClass.Spec.GetImageVersion())
 		}
-		versionName = lo.FromPtr(topImageVersionCandidate.Name)
 	}
 
-	selectedImageID := BuildImageID(publicGalleryURL, communityImageName, versionName)
-	if p.cm.HasChanged(key, selectedImageID) {
-		logging.FromContext(ctx).With("image-id", selectedImageID).Info("discovered new image id")
+	return "", fmt.Errorf("no compatible images found for instance type %s", instanceType.Name)
+}
+
+// GetLatestImageID returns the latest image ID for the given image version and image specification
+func (p *Provider) GetLatestImageID(ctx context.Context, defaultImage DefaultImageOutput, imageVersion string) (string, error) {
+	// Managed Karpenter will use the AKS Managed Shared Image Galleries
+	if options.FromContext(ctx).ManagedKarpenter {
+		return p.getImageIDSIG(ctx, defaultImage, imageVersion)
 	}
-	p.imageCache.Set(key, selectedImageID, imageExpirationInterval)
-	return selectedImageID, nil
+	// Self Hosted Karpenter will use the Community Image Galleries, which are public and have lower scaling limits
+	return p.getImageIDCIG(ctx, defaultImage.PublicGalleryURL, defaultImage.ImageDefinition, imageVersion)
 }
 
-func BuildImageID(publicGalleryURL, communityImageName, imageVersion string) string {
-	return fmt.Sprintf(imageIDFormat, publicGalleryURL, communityImageName, imageVersion)
+// getImageIDSIG retrieves the shared image gallery image URL in the format:
+// /subscriptions/{subscriptionID}/resourceGroups/{resourceGroup}/providers/Microsoft.Compute/galleries/{galleryName}/images/{imageDefinition}/versions/{imageVersion}.
+// If imageVersion is an empty string, it fetches the latest version of the image and caches it for imageExpirationInterval (3 days).
+func (p *Provider) getImageIDSIG(ctx context.Context, imgStub DefaultImageOutput, imageVersion string) (string, error) {
+	key := fmt.Sprintf(sharedImageGalleryImageIDFormat, options.FromContext(ctx).SIGSubscriptionID, imgStub.GalleryResourceGroup, imgStub.GalleryName, imgStub.ImageDefinition, imageVersion)
+	if imageID, ok := p.imageCache.Get(key); ok {
+		return imageID.(string), nil
+	}
+	if imageVersion == autoUpgradeMode {
+		versions, err := p.NodeImageVersions.List(ctx, p.location, p.subscription)
+		if err != nil {
+			return "", err
+		}
+		for _, version := range versions.Values {
+			imageID := fmt.Sprintf(sharedImageGalleryImageIDFormat, options.FromContext(ctx).SIGSubscriptionID, imgStub.GalleryResourceGroup, imgStub.GalleryName, version.SKU, version.Version)
+			cacheKey := fmt.Sprintf(sharedImageGalleryImageIDFormat, options.FromContext(ctx).SIGSubscriptionID, imgStub.GalleryResourceGroup, imgStub.GalleryName, version.SKU, autoUpgradeMode)
+			p.imageCache.Set(
+				cacheKey,
+				imageID, imageExpirationInterval)
+		}
+		// return the latest version of the image from the cache after we have caached all of the imageDefinitions
+		if imageID, ok := p.imageCache.Get(key); ok {
```

Comment:
What guarantees this key will be available?

Reply:
We iterate through all the  image definitions + image versions from ListNodeImageVersions call before this section and set them in the cache.  If we can't find a given image definition. 

```go
		for _, version := range versions.Values {
			imageID := fmt.Sprintf(sharedImageGalleryImageIDFormat, options.FromContext(ctx).SIGSubscriptionID, imgStub.GalleryResourceGroup, imgStub.GalleryName, version.SKU, version.Version)
			cacheKey := fmt.Sprintf(sharedImageGalleryImageIDFormat, options.FromContext(ctx).SIGSubscriptionID, imgStub.GalleryResourceGroup, imgStub.GalleryName, version.SKU, autoUpgradeMode)
			p.imageCache.Set(
				cacheKey,
				imageID, imageExpirationInterval)
		}

```